pkg/cmd/backuprestore/backupserver_test.go

```diff
@@ -0,0 +1,163 @@
+package backuprestore
+
+import (
+	"context"
+	"errors"
+	"github.com/robfig/cron"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+const validSchedule = "* * * * *"
+
+func TestBackupServer_Validate(t *testing.T) {
+	testCases := []struct {
+		name         string
+		backupServer backupServer
+		expErr       error
+	}{
+		{
+			"BackupServer is disabled",
+			backupServer{enabled: false},
+			nil,
+		},
+		{
+			"BackupServer is disabled and invalid schedule",
+			backupServer{
+				enabled:  false,
+				schedule: "invalid schedule",
+			},
+			nil,
+		},
+		{
+			"BackupServer is disabled and invalid schedule and invalid backup directory",
+			backupServer{
+				enabled:  false,
+				schedule: "invalid schedule",
+				backupOptions: backupOptions{
+					backupDir: "",
+				},
+			},
+			nil,
+		},
+		{
+			"BackupServer is enabled",
+			backupServer{
+				enabled: true,
+			},
+			errors.New("error parsing backup schedule : empty spec string"),
+		},
+		{
+			"BackupServer is enabled and invalid schedule",
+			backupServer{
+				enabled:  true,
+				schedule: "invalid schedule",
+			},
+			errors.New("error parsing backup schedule invalid schedule"),
+		},
+		{
+			"BackupServer is enabled and valid schedule",
+			backupServer{
+				enabled:  true,
+				schedule: validSchedule,
+			},
+			nil,
+		},
+		{
+			"BackupServer is enabled and invalid schedule and invalid backup directory",
+			backupServer{
+				enabled:  true,
+				schedule: "invalid schedule",
+				backupOptions: backupOptions{
+					backupDir: "",
+				},
+			},
+			errors.New("error parsing backup schedule invalid schedule"),
+		},
+		{
+			"BackupServer is enabled and valid schedule and invalid backup directory",
+			backupServer{
+				enabled:  true,
+				schedule: validSchedule,
+				backupOptions: backupOptions{
+					backupDir: "",
+				},
+			},
+			nil,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actErr := tc.backupServer.Validate()
+			if tc.expErr != nil {
+				require.Contains(t, actErr.Error(), tc.expErr.Error())
+			} else {
+				require.Equal(t, tc.expErr, actErr)
+			}
+		})
+	}
+}
+
+func TestNewBackupServer_scheduleBackup(t *testing.T) {
+	srvr := &backupServer{
+		timeZone:      "",
+		enabled:       true,
+		backupOptions: backupOptions{},
+	}
+
+	testCases := []struct {
+		name       string
+		schedule   string
+		timeout    time.Duration
+		expBackups int
+		expErr     error
+	}{
+		{
+			name:       "valid schedule",
+			schedule:   validSchedule,
+			timeout:    time.Minute * 3,
+			expBackups: 3,
+			expErr:     nil,
+		},
+		{
+			name:       "invalid schedule",
+			schedule:   "invalid schedule",
+			timeout:    time.Minute * 3,
+			expBackups: 0,
+			expErr:     errors.New("Expected exactly 5 fields"),
+		},
```

Comment:
please also add a case for:
* runBackup returns an error and still runs backups after that
* runBackup takes longer than your schedule

for the latter we would simply skip an invocation, correct?

Reply:
> runBackup takes longer than your schedule

I added this test case https://github.com/openshift/cluster-etcd-operator/pull/1306/commits/610246f3d3abde32245e84cd6f1db6ea2aba4c51

- sleep 3 secs, simulate slow backup
- timeout is 8 secs
- expected 3 backups to be taken 

please correct me if i am wrong 


> runBackup returns an error and still runs backups after that

if backup returns error, the whole process exits 

```
err := b.scheduleBackup(cCtx, bck)
		if err != nil {
			return err
		}
```

Reply:
> if backup returns error, the whole process exits

should this be the desired behavior?

Reply:
Well, we could log the error and keep on running, but how to know if the `backup()` is no functioning ?!!

Reply:
let me put it the other way then, as an admin how would you notice if it were crash looping?

Reply:
- statusCondition 
- alert
- logs
- kubectl :D :D 

the only notification that does not require polling is `alert`

Reply:
OK let's leave it crashing for now, we can discuss later how we can improve this